src/Block/Service/AbstractMenuBlockService.php

```diff
@@ -0,0 +1,197 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\BlockBundle\Block\Service;
+
+use Knp\Menu\ItemInterface;
+use Sonata\BlockBundle\Block\BlockContextInterface;
+use Sonata\BlockBundle\Form\Mapper\FormMapper;
+use Sonata\BlockBundle\Model\BlockInterface;
+use Sonata\Form\Type\ImmutableArrayType;
+use Symfony\Component\Form\Extension\Core\Type\CheckboxType;
+use Symfony\Component\Form\Extension\Core\Type\ChoiceType;
+use Symfony\Component\Form\Extension\Core\Type\TextType;
+use Symfony\Component\Form\FormTypeInterface;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+use Twig\Environment;
+
+abstract class AbstractMenuBlockService extends AbstractBlockService
+{
+    public function __construct(Environment $twig)
+    {
+        parent::__construct($twig);
+    }
+
+    public function execute(BlockContextInterface $blockContext, ?Response $response = null): Response
+    {
+        $template = $blockContext->getTemplate();
+        \assert(null !== $template);
+
+        $responseSettings = [
+            'menu' => $this->getMenu($blockContext),
+            'menu_options' => $this->getMenuOptions($blockContext->getSettings()),
+            'block' => $blockContext->getBlock(),
+            'context' => $blockContext,
+        ];
+
+        // NEXT_MAJOR: remove
+        if ('private' === $blockContext->getSetting('cache_policy')) {
+            return $this->renderPrivateResponse($template, $responseSettings, $response);
+        }
+
+        return $this->renderResponse($template, $responseSettings, $response);
+    }
+
+    public function configureCreateForm(FormMapper $form, BlockInterface $block): void
+    {
+        $this->configureEditForm($form, $block);
+    }
+
+    public function configureEditForm(FormMapper $form, BlockInterface $block): void
+    {
+        $form->add('settings', ImmutableArrayType::class, [
+            'keys' => $this->getFormSettingsKeys(),
+            'translation_domain' => 'SonataBlockBundle',
+        ]);
+    }
+
+    public function configureSettings(OptionsResolver $resolver): void
+    {
+        $resolver->setDefaults([
+            'title' => '',
+            // NEXT_MAJOR: Remove.
+            'cache_policy' => 'public',
+            'template' => '@SonataBlock/Block/block_core_menu.html.twig',
+            'safe_labels' => false,
+            'current_class' => 'active',
+            'first_class' => false,
+            'last_class' => false,
+            'current_uri' => null,
```

Comment:
Current_uri doesn't seems to be used here, but only in SeoBundle. I propose to remove it (in next major).